xapian-letor/tests/api_letor.cc

```diff
@@ -94,26 +167,127 @@ DEFINE_TESTCASE(createfeaturevector, generated)
     TEST_EQUAL(fv[0].get_fcount(), 19);
     TEST_EQUAL(fv[1].get_fcount(), 19);
+    return true;
+}
+
+DEFINE_TESTCASE(createfeaturevectoronevector, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker2",
+				       db_index_one_document);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("tigers"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("tigers"), db);
+    TEST(fv.empty());
+    mset = enquire.get_mset(0, 10);
+    TEST(!mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("tigers"), db);
+    TEST_EQUAL(fv.size(), 1);
+    TEST_EQUAL(fv[0].get_fcount(), 19);
+    return true;
+}
+
+DEFINE_TESTCASE(createfeaturevectoronevector_wrongquery, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker3",
+				       db_index_one_document);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("llamas"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("llamas"), db);
+    TEST(fv.empty());
+
+    mset = enquire.get_mset(0, 10);
+    TEST(mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("llamas"), db);
+    TEST_EQUAL(fv.size(), 0);
+    return true;
+}
+
+DEFINE_TESTCASE(createfeaturevectorthree, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker4",
+				       db_index_three_documents);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("score"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("score"), db);
+    TEST(fv.empty());
+
+    mset = enquire.get_mset(0, 10);
+    TEST(!mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("score"), db);
+    TEST_EQUAL(fv.size(), 2);
+    TEST_EQUAL(fv[0].get_fcount(), 19);
+    TEST_EQUAL(fv[1].get_fcount(), 19);
+    return true;
+}
+
+
+DEFINE_TESTCASE(preparetrainingfileonedb, generated)
+{
+    string db_path = get_database_path("apitest_listnet_ranker1",
+				       db_index_one_document);
+    string data_directory = test_driver::get_srcdir() + "/testdata/";
+    string query = data_directory + "queryone.txt";
+    string qrel = data_directory + "qrelone.txt";
+    string training_data = data_directory + "training_data_one_document.txt";
+    Xapian::prepare_training_file(db_path, query, qrel, 10,
+				  "training_output_data_one_doc.txt");
+    ifstream if1(training_data);
+    ifstream if2("training_output_data_one_doc.txt");
+    string file1;
+    string file2;
+    while (getline(if1, file1)) {
+	TEST(getline(if2, file2));
+	istringstream iss1(file1);
+	istringstream iss2(file2);
+	string temp1;
+	string temp2;
+	int i = 0;
+	while ((iss1 >> temp1) && (iss2 >> temp2)) {
+	    if (i == 0 || i == 1 || i == 21) {
```

Comment:
It would be good to have a comment explaining why 0, 1 and 21 are special.  I'm somewhat familiar with the code and I'm struggling to think - someone new to the code would likely have no idea.

Reply:
0th, 1st and 21st string taken as input, are absolute strings, "1", "qid:20001" and "#docid=1" respectively . 
So we can just check them as normal strings, whereas the other double values would have to tested as TEST_DOUBLE()

Reply:
OK, so can you explain that in a comment for the benefit of people trying to understand this in the future without them having to go digging through comments on github PRs (assuming github even still exists then).

One of those people could well be you - I think most experienced developers could tell you a tale of failing to follow some code which is rather lacking in comments, then digging into the version control history (e.g with `git blame`) to find out who wrote it, only to discover they themselves wrote it years earlier.